File: framework/backend/go/k1s0-domain-event/subscriber.go
package domainevent

import "context"

// EventHandler processes domain events of a specific type.
type EventHandler interface {
	// EventType returns the event type this handler processes.
	EventType() string

	// Handle processes the event envelope.
	Handle(ctx context.Context, envelope *EventEnvelope) error
}

// EventSubscriber manages event subscriptions.
type EventSubscriber interface {
	// Subscribe registers a handler and returns a function to cancel the subscription.
	Subscribe(handler EventHandler) (cancel func(), err error)
}

File: framework/backend/python/packages/k1s0-domain-event/src/k1s0_domain_event/subscriber.py
"""Event handler and subscriber abstract base classes."""

from __future__ import annotations

from abc import ABC, abstractmethod
from collections.abc import Callable, Coroutine
from typing import Any

from k1s0_domain_event.envelope import EventEnvelope


class EventHandler(ABC):
    """Abstract base class for handling domain events of a specific type."""

    @abstractmethod
    def event_type(self) -> str:
        """Return the event type this handler processes."""

    @abstractmethod
    async def handle(self, envelope: EventEnvelope) -> None:
        """Handle a domain event envelope.

        Args:
            envelope: The event envelope to handle.
        """


class EventSubscriber(ABC):
    """Abstract base class for subscribing event handlers."""

    @abstractmethod
    async def subscribe(
        self, handler: EventHandler
    ) -> Callable[[], Coroutine[Any, Any, None]]:
        """Subscribe a handler for its declared event type.

        Args:
            handler: The event handler to subscribe.

        Returns:
            An async callable that, when awaited, cancels the subscription.

        Raises:
            SubscribeError: If the subscription could not be created.
        """
